java-checks/src/main/java/org/sonar/java/closeresource/CloseableVisitor.java

```diff
@@ -0,0 +1,297 @@
+      if (usesIgnoredCloseableAsArgument(((NewClassTree) expression).arguments())) {
+        return new CloseableState.Ignored(expression);
+      }
+      return new CloseableState.Open(expression);
+    }
+    // TODO SONARJAVA-1029 : Engine currently ignore closeables which are retrieved from method calls. Handle them as OPEN.
+    return new CloseableState.Ignored(expression);
+  }
+
+
+  private static boolean isNull(ExpressionTree expression) {
+    return expression == null || expression.is(Tree.Kind.NULL_LITERAL);
+  }
+
+  private  boolean shouldBeIgnored(Symbol symbol, @Nullable ExpressionTree expression) {
+    return shouldBeIgnored(symbol) || shouldBeIgnored(expression);
+  }
+
+  private boolean shouldBeIgnored(Symbol symbol) {
+    return isSymbolIgnored(symbol)|| symbol.isFinal()
+        || isIgnoredCloseableSubtype(symbol.type())
+        || isSubclassOfInputStreamOrOutputStreamWithoutClose(symbol.type());
+  }
+
+  private static boolean shouldBeIgnored(@Nullable ExpressionTree expression) {
+    return expression != null && isSubclassOfInputStreamOrOutputStreamWithoutClose(expression.symbolType());
+  }
+
+  private boolean usesIgnoredCloseableAsArgument(List<ExpressionTree> arguments) {
+    for (ExpressionTree argument : arguments) {
+      if (isNewClassWithIgnoredArguments(argument)) {
+        return true;
+      } else if (isMethodInvocationWithIgnoredArguments(argument)) {
+        return true;
+      } else if (useIgnoredCloseable(argument) || isSubclassOfInputStreamOrOutputStreamWithoutClose(argument.symbolType())) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private boolean isNewClassWithIgnoredArguments(ExpressionTree argument) {
+    return argument.is(Tree.Kind.NEW_CLASS) && usesIgnoredCloseableAsArgument(((NewClassTree) argument).arguments());
+  }
+
+  private boolean isMethodInvocationWithIgnoredArguments(ExpressionTree argument) {
+    return argument.is(Tree.Kind.METHOD_INVOCATION) && usesIgnoredCloseableAsArgument(((MethodInvocationTree) argument).arguments());
+  }
+
+  private boolean useIgnoredCloseable(ExpressionTree expression) {
+    if (expression.is(Tree.Kind.IDENTIFIER, Tree.Kind.MEMBER_SELECT)) {
+      IdentifierTree identifier;
+      if (expression.is(Tree.Kind.MEMBER_SELECT)) {
+        identifier = ((MemberSelectExpressionTree) expression).identifier();
+      } else {
+        identifier = (IdentifierTree) expression;
+      }
+      if (isIgnoredCloseable(identifier.symbol())) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private boolean isIgnoredCloseable(Symbol symbol) {
+    if (CloseableVisitor.isCloseableOrAutoCloseableSubtype(symbol.type()) && !symbol.owner().isMethodSymbol()) {
+      return true;
+    } else {
+      return isSymbolIgnored(symbol);
+    }
+  }
+
+  private boolean isSymbolIgnored(Symbol symbol) {
+    List<State> statesOf = executionState.getStatesOf(symbol);
+    for (State state : statesOf) {
```

Comment:
:new: ![CRITICAL](https://raw.githubusercontent.com/henryju/image-hosting/master/critical.png)Possible null pointer dereference in org.sonar.java.closeresource.CloseableVisitor.isSymbolIgnored(Symbol) due to return value of called method[![...](https://raw.githubusercontent.com/henryju/image-hosting/master/more.png)](http://nemo.sonarqube.org/coding_rules#rule_key=findbugs%3ANP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE)